### PR TITLE
Applications: Manage versions, Apps Module Refactoring + Coverage Tests

### LINKED ISSUE #336: Add test coverage for Apps
As a developer, I'd like to improve the test coverage for apps tab.

### LINKED ISSUE #602: Add support to switch the default version in the app registry
As a user, I'd like to have the option to switch the default version in the app registry.

**Notes:**
The backend supports the option to switch the "default" version for an app. The following Shell command allows it.
> dataflow:>app default --id source:time --version 1.0.1.RELEASE

**Acceptance:**
- Enable this only on Skipper mode
- Remove duplicate rows for the same app from the list page (it shows multiple because of the different versions; we only want to show the "default" version)
- Design UI/UX to list all the available versions of each application
- The current app-list page is the good candidate as to where we would like to embed the versions
- This option is applicable to all the rows, so it'd be nice to consider how to reduce noise in each row

### PR DESCRIPTION
**Pull request by oodamien**
**Resolves #602:**
- manage versions application (skipper integration)
- Application Details: display properties of a selected version
- Apps: refactoring (workaround spring-cloud/spring-cloud-dataflow/#1871)

**Resolves #336:**
- Apps Bulk Import: refactoring using Form Reactive
- Apps Register: refactoring using Form Reactive
- Apps Details: added tests
- Application Details: tests
- Apps List: refactoring, tests
- Mock App Objects

**Global:**
- Added service/components/pipes: `ConfirmService`, `SortComponent`, `Pipe`...
- UI styles: quick improvements
- List: Refactoring: search/sort/pagination/state

**Other:**
- `SharedAppService`: fix type variable
- `Page<T>`: change Filter property (from string to object)

**Nb:**
- Task App can be extend the Apps List
- Applications list can reload data without lost the context (filters, page, selected items...)

**Todo:**
- Remove the context attributes from the model AppRegistration (force, selected Item)

### CODE REVIEW

**Comment by oodamien** [comment]
Refactoring Apps List: 
- Separate state/context and store
- Improve search experience
- Persist context
- Ready to be extended (for task applications)

**Comment by ghillert** [comment]
@oodamien Please rebase with master - there are currently some merge conflict.

File: ui/src/app/apps/apps.workaround.service.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ *
+ * @author Damien Vitrac
+ */
+@Injectable()
+export class AppsWorkaroundService {
+
+  public static cache = {
+    init: false,
+    invalid: true,
+    time: new Date().getTime(),
```
Comment (ghillert) [inline ui/src/app/apps/apps.workaround.service.ts]: I think we should use MomentJs for all our date-time needs. It is already used on the project

File: ui/src/app/apps/apps.workaround.service.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import {SharedAppsService} from '../shared/services/shared-apps.service';
+import {AppVersion} from '../shared/model/app-version';
+import {Page} from '../shared/model/page';
+import {SortParams} from '../shared/components/sort/sort.component';
+import {Subscriber} from 'rxjs/Subscriber';
+import {AppFilterParams, PaginationParams} from './components/apps.interface';
+
+/**
+ * Service class for the Apps module.
+ * Workaround, to be remove after resolve the related Spring Data Flow Server issues.
```
Comment (ghillert) [inline ui/src/app/apps/apps.workaround.service.ts]: Maybe add a few lines explaining the issues or add a Github issue ticket number here.

File: ui/src/app/shared/pipes/cesure.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  name: 'cesure'
+})
+export class CesurePipe implements PipeTransform {
+
+  transform(value: number | string, size: number = 40): string {
+    value = value.toString();
+    if (value.length <= size) {
+      return value;
+    }
+    return value.substr(0, size) + ' ...';
```
Comment (ghillert) [inline ui/src/app/shared/pipes/cesure.pipe.ts]: Should probably use the dedicated ellipsis `…` [`&hellip;`] character

File: ui/src/app/shared/pipes/cesure.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    return '... ' + value.substr(value.length - size, size);
+  }
+
+}
+
+
+@Pipe({
+  name: 'cesure'
+})
+export class CesurePipe implements PipeTransform {
```
Comment (ghillert) [inline ui/src/app/shared/pipes/cesure.pipe.ts]: Maybe rename this Pipe to `TruncatePipe` to make the meaning a bit clearer?

File: ui/src/app/shared/pipes/cesure.pipe.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import {Pipe, PipeTransform} from '@angular/core';
+
+@Pipe({
+  name: 'cesureLeft'
+})
+export class CesureLeftPipe implements PipeTransform {
```
Comment (ghillert) [inline ui/src/app/shared/pipes/cesure.pipe.ts]: Maybe rename this Pipe to TruncateLeftPipe to make the meaning a bit clearer?

File: ui/src/app/shared/pipes/cesure.pipe.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+  transform(value: number | string, size: number = 40): string {
+    value = value.toString();
+    if (value.length <= size) {
+      return value;
+    }
+    return '... ' + value.substr(value.length - size, size);
```
Comment (ghillert) [inline ui/src/app/shared/pipes/cesure.pipe.ts]: Should probably use the dedicated ellipsis … [&hellip;] character

Comment (ghillert) [inline ui/src/app/shared/pipes/cesure.pipe.ts]: I think this is fine for now BUT - Eventually we should be mindful of multi-byte characters. Definitely a follow-up issue - Maybe consider using something like https://github.com/bestiejs/punycode.js. This is a general issue across the app.

File: ui/src/app/shared/pipes/cesure.pipe.ts (hunk 5 above)
Comment (oodamien) [reply ui/src/app/shared/pipes/cesure.pipe.ts]: Merged with the first pipe (direction is a parameter)

Comment (oodamien) [reply ui/src/app/shared/pipes/cesure.pipe.ts]: + 'trail' (…) is a parameter

File: ui/src/app/apps/app-details/app-details.component.ts [hunk 7]
```diff
@@ -1,55 +1,195 @@
...
+import {ToastyService} from 'ng2-toasty';
+import {ApplicationType, DetailedAppRegistration} from '../../shared/model';
+import {Subscription} from 'rxjs/Subscription';
+import {SharedAboutService} from '../../shared/services/shared-about.service';
+import {AppRegistration} from '../../shared/model/app-registration.model';
+import {Observable} from 'rxjs/Observable';
+import {FeatureInfo} from '../../shared/model/about/feature-info.model';
+import {AppVersionsComponent} from '../app-versions/app-versions.component';
+import {BsModalService} from 'ngx-bootstrap';
+import {SortParams} from '../../shared/components/shared.interface';
```
Comment (ghillert) [inline ui/src/app/apps/app-details/app-details.component.ts]: import is missing for `combineLatest`, getting a runtime error. This works:

`import 'rxjs/add/observable/combineLatest';`

alternatively we may also embrace lettable operators and change as described here:

https://medium.com/@cartant/as-youve-used-it-combinelatest-isn-t-really-an-operator-it-s-an-observable-creator-bdab609dae58

**Comment by ghillert** [comment]
@oodamien Just a general thought: would it be better from a visual presentation to make the `labels` / `badges` on the App Registrations page the same width for all App Types?

image

I like the colors but feels my eyes is a bit jumping back and forth ;-)

File: ui/src/app/apps/apps/apps.component.html [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+              <span *ngIf="form.type == '' || form.type == null">
+                All types
+              </span>
+              <span *ngIf="form.type != '' && form.type!== null">
+                Type: <strong>{{ form.type }}</strong>
+              </span>
+              <span class="caret"></span>
+            </a>
+            <ul class="dropdown-menu dropdown-menu-right" *dropdownMenu="">
+              <li [class.active]="form.type == '' || form.type == null"><a (click)="form.type = ''">All type</a></li>
```
Comment (ghillert) [inline ui/src/app/apps/apps/apps.component.html]: Change `All type` to `All types`

image

Also, we may consider refactoring the `ApplicationType` domain class. Theoretically, it has all the relevant data to populate the drop-down data. BUT admittedly Typescript enums are "underwhelming".

We may consider using something like: https://github.com/LMFinney/ts-enums

File: ui/src/app/apps/apps/apps.component.html [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+          <button dropdownToggle type="button" class="btn btn-default btn-dropdown" data-toggle="dropdown"
+                  aria-haspopup="true"
+                  aria-expanded="false">
+            Actions&nbsp;
+            <span *ngIf="countSelected() == 1">(1)</span>
+            <span *ngIf="countSelected() > 1">({{ countSelected() }})</span>
+            <span class="caret"></span>
+          </button>
+          <ul *dropdownMenu class="dropdown-menu">
+            <li><a id="unregister-apps" href="#" (click)="unregisterAppsSelected()">Unregister application(s)</a></li>
```
Comment (ghillert) [inline ui/src/app/apps/apps/apps.component.html]: Maybe add  the number of messages to unregister and pluralize accordingly.

- `Unregister 1 application`
- `Unregister 4 applications`

image

**Comment by ghillert** [comment]
I wonder if all the icons should be on 1 line:

image

**Comment by ghillert** [comment]
Maybe you can also do some minor polishing in case of making the browser window smaller (aka mobile devices):

image

**Comment by ghillert** [comment]
We should standardize naming and change `Add Application(s)` to Register Application(s) or vice-versa.

image

Also, should we hide the `Add Application(s)` when no Apps are registered? (Since there is already a button visible)

**Comment by ghillert** [comment]
Strangely, when I hover over with the mouse, my pointer does not change to a pointed-finger?? 

image

File: ui/src/app/apps/apps.workaround.service.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+          .filter((a) => a != null)
+          .sort((a, b) => a.version < b.version ? -1 : 1);
+      });
+  }
+
+  private run(): Observable<any> {
+    return this.sharedAppsService
+      .getApps(new PageRequest(0, 1000), null, '', [{sort: 'name', order: 'ASC'}, {sort: 'type', order: 'ASC'}])
+      .map(page => {
+        console.log('DEBUG: HTTP REQUEST APPLICATION LIST');
```
Comment (ghillert) [inline ui/src/app/apps/apps.workaround.service.ts]: Remove debug statement

**Comment by ghillert** [comment]
Clicking on the version causes an error:

image

```
ERROR Error: StaticInjectorError(AppModule)[AppVersionsComponent -> ConfirmService]: 
  StaticInjectorError(Platform: core)[AppVersionsComponent -> ConfirmService]: 
    NullInjectorError: No provider for ConfirmService!
    at _NullInjector.get (core.js:994)
    at resolveToken (core.js:1292)
    at tryResolveToken (core.js:1234)
    at StaticInjector.get (core.js:1102)
    at resolveToken (core.js:1292)
    at tryResolveToken (core.js:1234)
    at StaticInjector.get (core.js:1102)
    at resolveNgModuleDep (core.js:10837)
    at NgModuleRef_.get (core.js:12070)
    at resolveDep (core.js:12560)
```

File: ui/src/app/shared/components/sort/sort.component.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
+import {Component, Output, EventEmitter, Input, ChangeDetectionStrategy, OnChanges, SimpleChanges} from '@angular/core';
+import {OrderParams, SortParams} from '../shared.interface';
+
+/**
+ * Sort
```
Comment (ghillert) [inline ui/src/app/shared/components/sort/sort.component.ts]: Add maybe a few more words

File: ui/src/app/apps/apps/apps.component.html (hunk 8 above)
Comment (oodamien) [reply ui/src/app/apps/apps/apps.component.html]: Hard coded - Have to be update with a new component lib

**Comment by ghillert** [comment]
LGTM - merging

**Pull request merged**